working/0546-patterns/patterns-feature-specification.md

```diff
@@ -144,953 +144,1073 @@ double calculateArea(Shape shape) =>
+patternFields         ::= patternField ( ',' patternField )* ','?
+patternField          ::= ( identifier? ':' )? pattern
 ```
 
-**TODO: Can type patterns have bounds?**
-
-The `typeOrBinder` rule is similar to the existing `type` grammar rule, but also
-allows `final` followed by an identifier to declare a type variable. It allows
-this at the top level of the rule and anywhere a type argument may appear inside
-a nested type. For example:
+A record pattern matches a record object destructures
+its fields. If the value isn't a record with the same shape as the pattern, then
+the match fails. Otherwise, the field subpatterns are matched against the
+corresponding fields in the record.
 
-```dart
-switch (object) {
-  case List<final E>: ...
-  case Map<String, final V>: ...
-  case Set<List<(final A, b: final B)>>: ...
-}
-```
+Field subpatterns can be in one of three forms:
 
-**TODO: Do we want to support function types? If so, how do we handle
-first-class generic function types?**
+*   A bare `pattern` destructures the corresponding positional field from the
+    record and matches it against `pattern`.
 
-#### List binder
+*   A `identifier: pattern` destructures the named field with the name
+    `identifier` and matches it against `pattern`.
 
-A list binder extracts elements by position from objects that implement `List`.
+*   A `: pattern` is a named field with the name omitted. When destructuring
+    named fields, it's very common to want to bind the resulting value to a
+    variable with the same name. As a convenience, the identifier can be omitted
+    on a named field. In that case, the name is inferred from `pattern`. The
+    subpattern must be a variable pattern or cast pattern, which may be wrapped
+    in any number of null-check or null-assert patterns.
 
-```
-listBinder ::= ('<' typeOrBinder '>' )? '[' binders ']'
-```
+    The field name is then inferred from the name in the variable or cast
+    pattern. These pairs of patterns are each equivalent:
 
-**TODO: Allow a `...` element in order to match suffixes or ignore extra
-elements. Allow capturing the rest in a variable.**
+    ```dart
+    // Variable:
+    (untyped: untyped, typed: int typed)
+    (:untyped, :int typed)
 
-#### Map binder
+    // Null-check and null assert:
+    (checked: checked?, asserted: asserted!)
+    (:checked?, :asserted!)
 
-A map binder access values by key from objects that implement `Map`.
+    // Cast:
+    (field: field as int)
+    (:field as int)
+    ```
 
-```
-mapBinder ::= mapTypeArguments? '{' mapBinderEntries '}'
+**TODO: Allow a `...` element in order to ignore some positional fields while
+capturing the suffix.**
 
-mapTypeArguments ::= '<' typeOrBinder ',' typeOrBinder '>'
+### Extractor pattern
 
-mapBinderEntries ::= mapBinderEntry ( ',' mapBinderEntry )* ','?
-mapBinderEntry ::= expression ':' binder
+```
+extractorPattern ::= extractorName typeArguments? '(' patternFields? ')'
```

Comment:
If you want to be able to match enum values, not just types, then you need to be able to match `ClassName<typeArgs>.name`.

So maybe `extractorName typeArguments? ('.' id)?` instead, to account for the enum element name?